src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                    {
+                        case SyntaxKind.MethodDeclaration:
+                        case SyntaxKind.LocalFunctionStatement:
+                        case SyntaxKind.PropertyDeclaration:
+                        case SyntaxKind.AnonymousMethodExpression:
+                            return node;
+                    }
+
+                    node = node.Parent;
+                }
+
+                return null;
+            }
+
+            private bool TryCreateStackFromQueryExpression(out QueryExpressionProcessingInfo queryExpressionProcessingInfo)
+            {
+                queryExpressionProcessingInfo = new QueryExpressionProcessingInfo(_source.FromClause);
+                return TryProcessQueryBody(_source.Body, queryExpressionProcessingInfo);
+            }
+
+            private StatementSyntax[] GenerateStatements(
+                Func<ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod,
+                QueryExpressionProcessingInfo queryExpressionProcessingInfo)
+            {
+                StatementSyntax statement = default;
+                var stack = queryExpressionProcessingInfo.Stack;
+                // Executes syntax building methods from bottom to the top of the tree.
+                // Process last clause
+                if (stack.Any())
+                {
+                    var node = stack.Pop();
+                    if (node is SelectClauseSyntax selectClause)
+                    {
+                        statement = WrapWithBlock(leafExpressionCreationMethod(selectClause.Expression));
+                    }
+                    else
+                    {
+                        throw new ArgumentException("Last node must me the select clause");
+                    }
+                }
+
+                // Process all other clauses
+                List<StatementSyntax> statements = new List<StatementSyntax>();
+                while (stack.Any())
+                {
+                    statement = ProcessClause(stack.Pop(), statement, out StatementSyntax extraStatement);
+                    if (extraStatement != null)
+                    {
+                        statements.Add(extraStatement);
+                    }
+                }
+
+                // The stack was processed in the reverse order, but the extra statements should be provided in the direct order.
+                statements.Reverse();
+                statements.Add(statement.WithAdditionalAnnotations(Simplifier.Annotation));
+                return statements.ToArray();
+            }
+
+            // TODO remove this method: should be split into 2 or refactored
+            private bool TryConvertQueryExpression(
+                Func<ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod,
+                out StatementSyntax[] statements)
+            {
+                statements = default;
+                if (TryCreateStackFromQueryExpression(out var stack))
```

Comment:
>TryCreateStackFromQueryExpression(out var stack) [](start = 20, length = 48)

It looks like we already calculated the stack and only need to thread it through the callers

Reply:
sure. I already remove this method.

---
In reply to: [178404166](https://github.com/dotnet/roslyn/pull/25362#discussion_r178404166) [](ancestors = 178404166)